File: 2439-minimize-maximum-of-array/2439-minimize-maximum-of-array.py
class Solution:
    def ceil(self, n: float) -> int:
        if n == int(n):
            return int(n)
        return int(n) + 1

    # flow goes to the left, max of all the averages
    def minimizeArrayValue(self, nums: List[int]) -> int:
        res = nums[0]
        running_sum = nums[0]
        for i in range(1, len(nums)):
            running_sum += nums[i]
            res = max(res, self.ceil(running_sum / (i+1)))
        return res

    # WRONG APPROACH
    # best flow is gonna spread the mean for each increasing sequence
    # => calculate all the means and return the max
    """
    def minimizeArrayValue(self, nums: List[int]) -> int:
        current_max = -sys.maxsize
        i = 1
        n = len(nums)
        running_avg = nums[0]
        running_count = 1
        while i < n:
            if nums[i] >= nums[i-1]:
                running_count += 1
                running_avg = running_avg + (nums[i] - running_avg) / running_count
            else:
                current_max = max(current_max, running_avg)
                running_avg = nums[i]
                running_count = 1
            i += 1
        current_max = max(current_max, running_avg)
        if current_max == int(current_max):
            return int(current_max)
        return int(current_max) + 1
    """

File: 2439-minimize-maximum-of-array/2439-minimize-maximum-of-array.rs
use std::cmp;

impl Solution {
    pub fn minimize_array_value(nums: Vec<i32>) -> i32 {
        let mut res: i32 = nums[0];
        let mut running_mean: f64 = nums[0] as f64;
        let mut running_count: f64 = 1 as f64;
        for i in 1..(nums.len()) {
            running_count = running_count + (1 as f64);
            running_mean = running_mean + ((nums[i as usize] as f64) - running_mean) / running_count;
            res = cmp::max(res, (running_mean.ceil()) as i32);
        }
        return res;
    }
}
